In [1]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import statsmodels.api as sm

In [3]:
# Importing the dfs
file_path1 = r"C:\Users\akash\Desktop\BootCamp\Project_2\Statistical-Modelling-with-Python-main\data\bike_station_info.csv"
file_path2 = r"C:\Users\akash\Desktop\BootCamp\Project_2\Statistical-Modelling-with-Python-main\data\poi_info.csv"
file_path3 = r"C:\Users\akash\Desktop\BootCamp\Project_2\Statistical-Modelling-with-Python-main\data\yelp_api.csv"
file_path4 = r"C:\Users\akash\Desktop\BootCamp\Project_2\Statistical-Modelling-with-Python-main\data\foursquare_api.csv"
file_path5 = r"C:\Users\akash\Desktop\BootCamp\Project_2\Statistical-Modelling-with-Python-main\data\bike_poi_info.csv"

# Read the CSV file into a DataFrame
city_bike_df = pd.read_csv(file_path1)
poi_df = pd.read_csv(file_path2)
yelp_df = pd.read_csv(file_path3)
foursquare_df = pd.read_csv(file_path4)
bike_poi_df = pd.read_csv(file_path5)

### Build a regression model.

In [5]:
bike_poi_df.head()

station_id  latitude_x  longitude_x  slots  \
0  b6903d788830595e054121e462598381   48.428023   -71.062506      7   
1  3aa299f4d3048ef94737bbf5fbc03825   48.404457   -71.058208      7   
2  0b1c0c48e94fae24b1a38f4559a5985a   48.420480   -71.053380      7   
3  f450af93c8d0f5ae5de8ea7cb1aafe70   48.337230   -70.879564      7   
4  7ada5d893c40c9a46a54f771a392db6f   48.425119   -71.051336      7   

   empty_slots  free_bikes                          name  latitude_y  \
0            4           2  Le Légendaire Restaurant Bar   48.410643   
1            1           6                       Boefish   48.405576   
2            5           2                  Passion Café   48.409290   
3            5           2          Rôtisserie St-Hubert   48.413318   
4            5           2              Ben & Florentine   48.403481   

   longitude_y  yelp_rating  yelp_review  
0   -71.053362          4.0            6  
1   -71.057046          4.5            5  
2   -71.054947          4.0            5  
3   -71.049191          2.5            4  
4   -71.054605          2.5            6

In [6]:
# First I set our dependent and independant variables
# Load the dataset
data = bike_poi_df

# Prepare the feature matrix X and target variable y
X = data[['slots', 'free_bikes']]
y = data['empty_slots']

# Add a constant term to the feature matrix
X = sm.add_constant(X)

# Fit the regression model
model = sm.OLS(y, X)
results = model.fit()

# Print the model summary
print(results.summary())

                            OLS Regression Results                            
Dep. Variable:            empty_slots   R-squared:                       0.961
Model:                            OLS   Adj. R-squared:                  0.953
Method:                 Least Squares   F-statistic:                     122.2
Date:                Sun, 02 Jul 2023   Prob (F-statistic):           0.000105
Time:                        20:04:50   Log-Likelihood:                -2.5204
No. Observations:                   7   AIC:                             9.041
Df Residuals:                       5   BIC:                             8.933
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
slots          0.9699      0.039     24.735      0.0

d:\Software\Anaconda\envs\bootcamp_env_1\lib\site-packages\statsmodels\stats\stattools.py:74: ValueWarning: omni_normtest is not valid with less than 8 observations; 7 samples were given.
  warn("omni_normtest is not valid with less than 8 observations; %i "


Run the model a couple of time to get great fit.

#### Provide model output and an interpretation of the results. 

- The model has a high R-squared value of 0.961, indicating a strong fit between the independent variables and the dependent variable.
- The coefficients show that 'slots' has a positive effect on the number of empty slots, while 'free_bikes' has a negative effect.
- Both 'slots' and 'free_bikes' have statistically significant effects on the number of empty slots, as indicated by their low p-values.
- The F-statistic of 122.2 suggests that the model as a whole is statistically significant.
- The adjusted R-squared value of 0.953 indicates a good fit while considering the number of variables.
- The standard errors provide a measure of uncertainty in the estimated coefficients.
- The model suggests that 'slots' and 'free_bikes' are important predictors of the number of empty slots in the given context.

[NOTE] The warning indicates that the statistical test being used, specifically the omnibus normality test, requires a minimum of 8 observations, but only 7 observations were provided in this case.

# Stretch

How can you turn the regression model into a classification model?

- To convert the regression model into a classification model, I can apply a threshold to the predicted values, By setting a threshold, I can categorize the data into different classes based on whether the predicted value exceeds the threshold or not.
- The threshold value is determined based on distance of the cycle to the stores or other pois or other specific requirements or characteristics of the problem at hand.
- This approach allows me to classify new data points into appropriate classes based on their predicted values, making it suitable for classification tasks.